### PR TITLE
Reduce allocations in MakePath() method

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

In module loading scenario PerfView shows a lot of allocations of StringBuilder objects
image

The fix is to use System.IO.Path.Join() method.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/namespaces/NavigationProviderBase.cs [hunk 1]
```diff
@@ -375,44 +375,13 @@ protected string MakePath(string parent, string child, bool childIsLeaf)
```
Comment (PoshChan) [inline src/System.Management.Automation/namespaces/NavigationProviderBase.cs]: @iSazonov, your last commit had 123 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify Directory + Whatif
```powershell
Expected an exception, with FullyQualifiedErrorId 'PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand' to be thrown, but the FullyQualifiedErrorId was 'System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemCommand'. from D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1:900 char:15
    +             { Get-Item -Path $testDir -ErrorAction Stop } |
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 901
901:                 Should -Throw -ErrorId "PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand"
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify File + Value
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 918
918:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Force
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 926
926:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Path Wildcard
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 931
931:             $result.Name | Should -BeExactly $testFile2
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Include
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 936
936:             $result.Name | Should -BeExactly $testFile2
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/NavigationProviderBase.cs]: @iSazonov, your last commit had 130 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify Directory + Whatif
```powershell
Expected an exception, with FullyQualifiedErrorId 'PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand' to be thrown, but the FullyQualifiedErrorId was 'System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemCommand'. from /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:900 char:15
    +             { Get-Item -Path $testDir -ErrorAction Stop } |
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 901
901:                 Should -Throw -ErrorId "PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand"
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify File + Value
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 918
918:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Force
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 926
926:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Path Wildcard
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 931
931:             $result.Name | Should -BeExactly $testFile2
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Include
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 936
936:             $result.Name | Should -BeExactly $testFile2
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/NavigationProviderBase.cs]: @iSazonov, your last commit had 130 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify Directory + Whatif
```powershell
Expected an exception, with FullyQualifiedErrorId 'PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand' to be thrown, but the FullyQualifiedErrorId was 'System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemCommand'. from /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:900 char:15
    +             { Get-Item -Path $testDir -ErrorAction Stop } |
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 901
901:                 Should -Throw -ErrorId "PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand"
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify File + Value
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 918
918:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Force
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 926
926:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Path Wildcard
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 931
931:             $result.Name | Should -BeExactly $testFile2
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Include
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 936
936:             $result.Name | Should -BeExactly $testFile2
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/NavigationProviderBase.cs]: @iSazonov, your last commit had 130 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify Directory + Whatif
```powershell
Expected an exception, with FullyQualifiedErrorId 'PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand' to be thrown, but the FullyQualifiedErrorId was 'System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemCommand'. from /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:900 char:15
    +             { Get-Item -Path $testDir -ErrorAction Stop } |
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 901
901:                 Should -Throw -ErrorId "PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand"
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify File + Value
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 918
918:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Force
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 926
926:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Path Wildcard
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 931
931:             $result.Name | Should -BeExactly $testFile2
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Include
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 936
936:             $result.Name | Should -BeExactly $testFile2
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/NavigationProviderBase.cs]: @iSazonov, your last commit had 123 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify Directory + Whatif
```powershell
Expected an exception, with FullyQualifiedErrorId 'PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand' to be thrown, but the FullyQualifiedErrorId was 'System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemCommand'. from D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1:900 char:15
    +             { Get-Item -Path $testDir -ErrorAction Stop } |
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 901
901:                 Should -Throw -ErrorId "PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand"
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify File + Value
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 918
918:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Force
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 926
926:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Path Wildcard
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 931
931:             $result.Name | Should -BeExactly $testFile2
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Include
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\FileSystem.Tests.ps1: line 936
936:             $result.Name | Should -BeExactly $testFile2
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/NavigationProviderBase.cs]: @iSazonov, your last commit had 130 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify Directory + Whatif
```powershell
Expected an exception, with FullyQualifiedErrorId 'PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand' to be thrown, but the FullyQualifiedErrorId was 'System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemCommand'. from /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:900 char:15
    +             { Get-Item -Path $testDir -ErrorAction Stop } |
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 901
901:                 Should -Throw -ErrorId "PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand"
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate New-Item parameters.Verify File + Value
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 918
918:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Force
```powershell
Expected exactly 'testFile.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 926
926:             $result.Name | Should -BeExactly $testFile
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Path Wildcard
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 931
931:             $result.Name | Should -BeExactly $testFile2
```
Extended FileSystem Item/Content Cmdlet Provider Tests.Valdiate Get-Item parameters.Verify Include
```powershell
Expected exactly 'testFile2.txt', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 936
936:             $result.Name | Should -BeExactly $testFile2
```

**Comment by iSazonov** [comment]
@daxian-dbw The PR is ready to review. I removed all obvious comments so diff shows over 50 changes but really there is only 2 functional change - use Path.Join and a change in one condition to fix edge case (child is one slash string).

**Comment by iSazonov** [comment]
@SteveL-MSFT @daxian-dbw Could you please review?

**Review by daxian-dbw — APPROVED** [review]

**Comment by iSazonov** [comment]
@adityapatwardhan Could you please merge?

**Comment by iSazonov** [comment]
@daxian-dbw @adityapatwardhan Can we get a progress here?

**Comment by daxian-dbw** [comment]
@adityapatwardhan Could you please take a look?

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**